Add support for custom HTTP headers

jvff opened the pull request:
HttpHandle can be configured to include custom HTTP headers in the RPC
requests it sends.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)
<!-- Reviewable:end -->

jvff commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/src/lib.rs, line 42 at r1](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7PeV-rflvU8ixvsREn:-L7PeV-uoQxCFzESuFdd:b-svv0fk) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/0ce17ba5d40cdb080adc9703efda2fbda6ca53e5/http/src/lib.rs#L42)):*
> ```Rust
> //!
> //! ```rust,no_run
> //! #[macro_use] extern crate jsonrpc_client_core;
> ```

Oops. Rustfmt bug. Will force push a fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
This turned out way simpler than I had expected. Really sweet! :+1: Would be awesome if you could test to use an IP in the `uri` but then set a custom `Host` header and see if it works.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/src/lib.rs, line 340 at r2](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7PgEe6Ak6JJlcJZxnN:-L7PgEe7CPOW_N-g6ZCV:b-3fdtw5) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/dbd39c4addc0cd4a26f1a539fc9f3eeafd3962ca/http/src/lib.rs#L340)):*
> ```Rust
>     fn create_request(&self, body: Vec<u8>) -> Request {
>         let mut request = hyper::Request::new(hyper::Method::Post, self.uri.clone());
>         request.headers_mut().extend(self.headers.iter());
> ```

Have you verified what happens if `self.headers` contains a field already existing on the request? The documentation for `extend` says it will overwrite unless the collection supports multiple keys with the same value. Not sure how `Request` is implemented.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-:-L7Pge3d0bfnLeN6_Z9e:b-fo14mg)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 3 files at r3.
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

jvff commented:
I wrote some tests, but they might be a little dirty/hacky. Any suggestions for improvements? Or leave this as-is for now?

---

Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/src/lib.rs, line 340 at r2](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7PgEe6Ak6JJlcJZxnN:-L7R4QUuy0T3xZEb6hj3:b-nx3vzj) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/dbd39c4addc0cd4a26f1a539fc9f3eeafd3962ca/http/src/lib.rs#L340)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Have you verified what happens if `self.headers` contains a field already existing on the request? The documentation for `extend` says it will overwrite unless the collection supports multiple keys with the same value. Not sure how `Request` is implemented.
</blockquote></details>

I tested setting the same header with different values (i.e., extending with a header set to two different values), but not extending over an existing header.

It shouldn't happen, because IIUC there are no existing headers, and `Content-Type` and `Content-Length` are set afterwards, overwriting any user-configured value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-:-L7R5BhiN10giOGxbthU:b-cq1xbm)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[http/src/lib.rs, line 340 at r2](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7PgEe6Ak6JJlcJZxnN:-L7R1kWqh04uizviOcG3:bj2qymu) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/dbd39c4addc0cd4a26f1a539fc9f3eeafd3962ca/http/src/lib.rs#L340)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Have you verified what happens if `self.headers` contains a field already existing on the request? The documentation for `extend` says it will overwrite unless the collection supports multiple keys with the same value. Not sure how `Request` is implemented.
</blockquote></details>

Would be nice if it was actually possible to override `ContentType` and `ContentLength` as well. Right? Otherwise it's impossible to override them even if one would like to. Can probably be done just by moving this line to just before `request.set_body(body);`.

---

*[http/tests/custom_headers.rs, line 82 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R17tFx3uvgF5Tr4J7:-L7R17tGRaZY7ExL_ox8:b-1x0ky6) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L82)):*
> ```Rust
> }
> 
> fn test_custom_headers<H, C>(port: u16, set_headers: H, check_headers: C)
> ```

I'm guessing `C` stands for `check`? Then why is `H` representing `set`? :wink: 

---

*[http/tests/custom_headers.rs, line 102 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R4xw5F0Qm2Nc_e0Sy:-L7R4xw5F0Qm2Nc_e0Sz:bvpqdky) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L102)):*
> ```Rust
> 
>     let send = transport_handle.send(Vec::new()).map_err(mem::drop);
>     let future = send.and_then(|_| rx.map_err(mem::drop));
> ```

Would you not just chain the `.and_then` directly onto the first expression? The names `send` and `future` did not immediately make me understand what they were, and by chaining everything together you only have to name one variable.

---

*[http/tests/custom_headers.rs, line 106 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R4gYMngVl7JP2Y6z3:-L7R4gYMngVl7JP2Y6z4:b-k5nntp) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L106)):*
> ```Rust
>     let result = reactor.run(future).unwrap();
> 
>     assert_eq!(result, true);
> ```

`assert!(result)`?

---

*[http/tests/custom_headers.rs, line 175 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R2uMxLSsjKqSLuEJ3:-L7R2uMxLSsjKqSLuEJ4:b84riw1) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L175)):*
> ```Rust
> fn spawn_server<F>(port: u16, check: F) -> ServerShutdownFlag
> where
>     F: 'static + FnOnce(Request) + Send,
> ```

This is just a stylistic question, but I always put the lifetime bound last. I usually do
1) trait bounds,
2) send/sync bounds (which are also traits, albeit rather special ones),
3) lifetime bounds.

---

*[http/tests/custom_headers.rs, line 181 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R5ixgY76tHCWSdoY0:-L7R5ixgY76tHCWSdoY1:b-rpmnic) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L181)):*
> ```Rust
>     thread::spawn(move || {
>         let service = CheckService::new(check);
>         let address = SocketAddr::new("127.0.0.1".parse().unwrap(), port);
> ```

You have no guarantee that the ports you use will be free. Introducing an element of undesired randomness here. If you instead use port `0` the OS will automatically allocate a random free port. That is not very likely to ever fail.

After server creation you can get the port out with `Server::local_addr`.

Another benefit of this is that you will have to send the port out of this thread. Which will force you to create a channel back into the parent thread. Thus you can remove the 100 ms sleep, instead you know as soon as you have a message on this port-channel the server is up and running :)

---

*[http/tests/custom_headers.rs, line 186 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7R3krHlnA3naJuUnhQ:b-wtz8n1) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
> ```Rust
>             .unwrap();
> 
>         server.run_until(rx.map_err(mem::drop)).unwrap();
> ```

Why do you map the error with `mem::drop`? That function returns `()` so would this not be identical to just `map_err(|_| ())` which IMO is easier to understand?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
I agree the tests became a bit complex. But I really like that you put effort into testing! This crate could use more integration tests in general. So if you get a nice base test rig up it will be easier to add tests later.

I have some suggestions for improvements in the code.

---

Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-:-L7R6Dk4O030mjB7Ry6a:b-jifida)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[http/src/lib.rs, line 340 at r2](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7PgEe6Ak6JJlcJZxnN:-L7TqS5cBJ6cBLyPmceU:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/dbd39c4addc0cd4a26f1a539fc9f3eeafd3962ca/http/src/lib.rs#L340)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Would be nice if it was actually possible to override `ContentType` and `ContentLength` as well. Right? Otherwise it's impossible to override them even if one would like to. Can probably be done just by moving this line to just before `request.set_body(body);`.
</blockquote></details>

Done.

---

*[http/tests/custom_headers.rs, line 82 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R17tFx3uvgF5Tr4J7:-L7TqbHQC136yhP2h0QJ:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L82)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I'm guessing `C` stands for `check`? Then why is `H` representing `set`? :wink: 
</blockquote></details>

Done.

---

*[http/tests/custom_headers.rs, line 102 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R4xw5F0Qm2Nc_e0Sy:-L7TrLla4oW143Vr5N36:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L102)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Would you not just chain the `.and_then` directly onto the first expression? The names `send` and `future` did not immediately make me understand what they were, and by chaining everything together you only have to name one variable.
</blockquote></details>

Done.

---

*[http/tests/custom_headers.rs, line 106 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R4gYMngVl7JP2Y6z3:-L7TrewvEXDPk-RWeGkK:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L106)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

`assert!(result)`?
</blockquote></details>

Done.

---

*[http/tests/custom_headers.rs, line 175 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R2uMxLSsjKqSLuEJ3:-L7TrsuXA60GDNvDTyKw:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L175)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

This is just a stylistic question, but I always put the lifetime bound last. I usually do
1) trait bounds,
2) send/sync bounds (which are also traits, albeit rather special ones),
3) lifetime bounds.
</blockquote></details>

Done.

---

*[http/tests/custom_headers.rs, line 181 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R5ixgY76tHCWSdoY0:-L7Ts-vg6EgJzJ8vH32I:b-qwa5j6) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L181)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

You have no guarantee that the ports you use will be free. Introducing an element of undesired randomness here. If you instead use port `0` the OS will automatically allocate a random free port. That is not very likely to ever fail.

After server creation you can get the port out with `Server::local_addr`.

Another benefit of this is that you will have to send the port out of this thread. Which will force you to create a channel back into the parent thread. Thus you can remove the 100 ms sleep, instead you know as soon as you have a message on this port-channel the server is up and running :)
</blockquote></details>

This is a really nice improvement! Done :)

---

*[http/tests/custom_headers.rs, line 186 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7TuRpv58ObmdNHYDeF:b-jnf60b) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Why do you map the error with `mem::drop`? That function returns `()` so would this not be identical to just `map_err(|_| ())` which IMO is easier to understand?
</blockquote></details>

`.map_err(|_| ())` is more consistent, but what I really wanted was `.drop_err()` :)
I feel tempted to write that extension trait just for these use cases.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/src/lib.rs, line 328 at r4](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7UJJTY8tDMH03qVb8H:-L7UJJTZ0QF1jZDccbT2:bl6gcpz) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/8af434a23530adbaa10be770c48019a6b27e93d7/http/src/lib.rs#L328)):*
> ```Rust
> 
> impl HttpHandle {
>     /// Configure a custom HTTP header for the requests.
> ```

Maybe change "the requests" to "all requests sent through this transport"?

---

*[http/src/lib.rs, line 331 at r4](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7UIbe7CxQ1EaRF6Crq:-L7UIbe7CxQ1EaRF6Crr:bw09pty) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/8af434a23530adbaa10be770c48019a6b27e93d7/http/src/lib.rs#L331)):*
> ```Rust
>     ///
>     /// The ContentType and ContentLength headers are initially set based on the request, but can
>     /// be manually changed.
> ```

These are not the only headers set by default, are they? I guess hyper will set for example the `Host` header as well? Which can also be overridden with this method. So maybe make it a bit more general?

I don't find it apparent how it's "manually changed".

How about "Replaces any header set by this library or by Hyper, such as the ContentType, ContentLength and Host headers."

---

*[http/tests/custom_headers.rs, line 186 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7ULgnC352Lfw2sameB:bycsk8i) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

`.map_err(|_| ())` is more consistent, but what I really wanted was `.drop_err()` :)
I feel tempted to write that extension trait just for these use cases.
</blockquote></details>

`.map_err(|_| ())` would be equal to a `.drop_err()` if I correctly understood what you wanted. Why would you want to create a whole new trait for such a simple operation? Why must the error be dropped? I don't see any part of the error holding any resource you wanted to free or anything.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[http/src/lib.rs, line 328 at r4](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7UJJTY8tDMH03qVb8H:-L7UMYp-79mlMG8yT2qu:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/8af434a23530adbaa10be770c48019a6b27e93d7/http/src/lib.rs#L328)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Maybe change "the requests" to "all requests sent through this transport"?
</blockquote></details>

Done.

---

*[http/src/lib.rs, line 331 at r4](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7UIbe7CxQ1EaRF6Crq:-L7UMcXc-IML2TFgVl6F:b-gv52fm) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/8af434a23530adbaa10be770c48019a6b27e93d7/http/src/lib.rs#L331)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

These are not the only headers set by default, are they? I guess hyper will set for example the `Host` header as well? Which can also be overridden with this method. So maybe make it a bit more general?

I don't find it apparent how it's "manually changed".

How about "Replaces any header set by this library or by Hyper, such as the ContentType, ContentLength and Host headers."
</blockquote></details>

That's a lot clearer, thanks :)

---

*[http/tests/custom_headers.rs, line 186 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7UMxqZ6QlDxVE2JhZo:betgnc4) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

`.map_err(|_| ())` would be equal to a `.drop_err()` if I correctly understood what you wanted. Why would you want to create a whole new trait for such a simple operation? Why must the error be dropped? I don't see any part of the error holding any resource you wanted to free or anything.
</blockquote></details>

In this case, the required error type for `run_until` is `()`, so I have to drop any error that may appear.

In the usages that occur before this one, the error doesn't need to be dropped, but the `AndThen` future requires the futures to have the same error type. Since I didn't need any of them for the test, I could just drop them and make the error type `()`.

The only reason to create the trait would be to write `.drop_err()` instead of `.map_err(|_| ())`, but I agree that it's added code for a marginal aesthetics change.

I gave it some more thought and I think it might be better to use `.map_err(|err| err.to_string())` for the code in the `AndThen`. What do you think?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/tests/custom_headers.rs, line 186 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7V45ib6fSwI-EX5Q5a:b-dneoew) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

In this case, the required error type for `run_until` is `()`, so I have to drop any error that may appear.

In the usages that occur before this one, the error doesn't need to be dropped, but the `AndThen` future requires the futures to have the same error type. Since I didn't need any of them for the test, I could just drop them and make the error type `()`.

The only reason to create the trait would be to write `.drop_err()` instead of `.map_err(|_| ())`, but I agree that it's added code for a marginal aesthetics change.

I gave it some more thought and I think it might be better to use `.map_err(|err| err.to_string())` for the code in the `AndThen`. What do you think?
</blockquote></details>

I don't think we use the word "drop" in the same way. Yes, your error type must be `()`, that does not mean your previous error must be dropped, as in deallocated. You only have to shape your futures error type type into becoming a `()`, which `map_err(|_| ())` will do. That `map_err` does also have the effect that the previous error is actually dropped, in the deallocation meaning of the word, since there are no references owning it after the mapping.

But on the other hand something like this (non working example):
```rust
let mut errors = vec![];
...
some_future.map_err(|e| {
    errors.push(e);
    ()
});
```
Would also satisfy the requirements for the error type to be `()` but it would not drop the old error. And something like
```rust
some_future.map(|_| String::from("The error type is now String"))
```
Would deallocate/drop the error, but it would not satisfy your required type of `()`. Just arguing that the deallocation of your error instance is not the exact property you are after, just that you want the type to be `()` when you are done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/tests/custom_headers.rs, line 186 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7V7-zCwpNB3rrnclCU:bqivk1u) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I don't think we use the word "drop" in the same way. Yes, your error type must be `()`, that does not mean your previous error must be dropped, as in deallocated. You only have to shape your futures error type type into becoming a `()`, which `map_err(|_| ())` will do. That `map_err` does also have the effect that the previous error is actually dropped, in the deallocation meaning of the word, since there are no references owning it after the mapping.

But on the other hand something like this (non working example):
```rust
let mut errors = vec![];
...
some_future.map_err(|e| {
    errors.push(e);
    ()
});
```
Would also satisfy the requirements for the error type to be `()` but it would not drop the old error. And something like
```rust
some_future.map(|_| String::from("The error type is now String"))
```
Would deallocate/drop the error, but it would not satisfy your required type of `()`. Just arguing that the deallocation of your error instance is not the exact property you are after, just that you want the type to be `()` when you are done.
</blockquote></details>

I don't see the need to convert the error into a string. The error can only happen if the other end of the channel has been closed, which only happens when we don't care about the server any longer. So what would we use the error for?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/tests/custom_headers.rs, line 186 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7VVWZbBRnVtBcyjfmu:b-tzkcl1) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I don't see the need to convert the error into a string. The error can only happen if the other end of the channel has been closed, which only happens when we don't care about the server any longer. So what would we use the error for?
</blockquote></details>

You're right, I was (ab)using the semantics of drop. What I really meant was "ignore error", but what you said about mapping it to be `()` now makes more sense to me.  In that case `.map_err(|_| ())` actually is a lot clearer about the meaning. Sorry for the misunderstanding.

Regarding placing the error in a `String`, I guess it's more of a auxiliary message if the error occurs. It would help little, but would save a few seconds to figure out if the problem was the server not sending the message or if it closed the channel without notifying the test result.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/tests/custom_headers.rs, line 186 at r3](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7VZvaplj4PDp04mpua:b-tafukr) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

You're right, I was (ab)using the semantics of drop. What I really meant was "ignore error", but what you said about mapping it to be `()` now makes more sense to me.  In that case `.map_err(|_| ())` actually is a lot clearer about the meaning. Sorry for the misunderstanding.

Regarding placing the error in a `String`, I guess it's more of a auxiliary message if the error occurs. It would help little, but would save a few seconds to figure out if the problem was the server not sending the message or if it closed the channel without notifying the test result.
</blockquote></details>

I'm not sure which `AndThen` you mean. If `shutdown_rx` returns the error type it's not really an error, it just means the sending half has been dropped. I think just keep it simple, let the server run until the channel is closed/send/whatever and then we are done, this is just test boilerplate code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 1 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/tests/custom_headers.rs, line 186 at r3](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7R3krG3XD66ZU7MddG:-L7VaPqrCDf0AiJCb7hC:b-g7gauh) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/f162f75300a0f933937aece08a55de2e3bd4f952/http/tests/custom_headers.rs#L186)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I'm not sure which `AndThen` you mean. If `shutdown_rx` returns the error type it's not really an error, it just means the sending half has been dropped. I think just keep it simple, let the server run until the channel is closed/send/whatever and then we are done, this is just test boilerplate code.
</blockquote></details>

Sorry, not this `AndThen`. I misused Reviewable's line comment feature :/

I'll open a discussion on the correct line.

---

*[http/tests/custom_headers.rs, line 125 at r4](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7VarVl0nt4zj7LMPG3:-L7VarVl0nt4zj7LMPG4:bqoihbn) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/8af434a23530adbaa10be770c48019a6b27e93d7/http/tests/custom_headers.rs#L125)):*
> ```Rust
>         .send(Vec::new())
>         .map_err(|_| ())
>         .and_then(|_| rx.map_err(|_| ()));
> ```

These are the two `map_err`s that I'm thinking about changing to `|err| err.to_string()`.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 1 files at r5.
Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[http/tests/custom_headers.rs, line 125 at r4](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7VarVl0nt4zj7LMPG3:-L7YOZ9je5v6OIQ1HEzm:b-aq2wj0) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/8af434a23530adbaa10be770c48019a6b27e93d7/http/tests/custom_headers.rs#L125)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

These are the two `map_err`s that I'm thinking about changing to `|err| err.to_string()`.
</blockquote></details>

Ah I see. Now I understand who/where any error would pop up. In the `reactor.run(..).unwrap()` unwrapping. Yeah, that makes sense. Do that if you still want to. Good if errors can spit out hints to what crashed indeed.

---

*[http/tests/custom_headers.rs, line 116 at r5](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7YO4RUqeDASh_LoHUZ:-L7YO4RVkzQDRrB3fo0f:b-yg4jtx) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L116)):*
> ```Rust
> 
>     let transport = HttpTransport::new().unwrap();
>     let uri = format!("http://localhost:{}", port);
> ```

`localhost` does not always work. It's a hostname as all others and if the `/etc/hosts` file or equivalent on other systems is not correctly set up then the computer can't reach localhost. I have actually experienced this on a number of computers. However `127.0.0.1` always works. And since you use the IP directly when setting up the server, it would be more consistent to do it here as well. Or do you want the hostname for some specific reason? Such as testing the `Host` field or such?

---

*[http/tests/custom_headers.rs, line 164 at r5](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7YThBpRvTVuiPQeCpG:-L7YThBpRvTVuiPQeCpH:b-f4448t) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L164)):*
> ```Rust
> }
> 
> pub struct ServerShutdownFlag(Option<Sender<()>>);
> ```

I have personally not seen "flag" used in this context I think. An object one keep track of in order to control another object somewhere I usually refer to as a "handle".

For me a flag is usually something that is sent in or returned out to indicate the activation of some feature.

---

*[http/tests/custom_headers.rs, line 168 at r5](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7YT9Mnhrs06-hMMnEJ:-L7YT9MoJMdwEvILnTnO:bwjrukb) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L168)):*
> ```Rust
> impl Drop for ServerShutdownFlag {
>     fn drop(&mut self) {
>         self.0.take().unwrap().send(()).unwrap();
> ```

Would it not work equally well without this `Drop` implementation? Since when `ServerShutdownFlag` is dropped the `Sender` will automatically be dropped and thus closed anyway. So then the `Receiver` will return in an error since the channel is closed. Both would make `run_until` return? Just thinking we can simplify the implementation a bit maybe.

---

*[http/tests/custom_headers.rs, line 180 at r5](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7YUlpg0dYuu8XKXOem:-L7YUlphtEXpPyWg9NMZ:b-9puzgb) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L180)):*
> ```Rust
> }
> 
> impl<S: Service> NewService for OneNewService<S> {
> ```

I have never used tokio services. Is this the simplest way to just listen for a single request on a http server? Seems like a lot of boilerplate to define two types implementing two traits just for that. But looking at the `Http` documentation it looks like it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[http/tests/custom_headers.rs, line 125 at r4](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7VarVl0nt4zj7LMPG3:-L7YqjbDEmpyYg7hQd11:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/8af434a23530adbaa10be770c48019a6b27e93d7/http/tests/custom_headers.rs#L125)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Ah I see. Now I understand who/where any error would pop up. In the `reactor.run(..).unwrap()` unwrapping. Yeah, that makes sense. Do that if you still want to. Good if errors can spit out hints to what crashed indeed.
</blockquote></details>

Done.

---

*[http/tests/custom_headers.rs, line 116 at r5](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7YO4RUqeDASh_LoHUZ:-L7Z3FH5217OcjVpUhOU:b1lj74b) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L116)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

`localhost` does not always work. It's a hostname as all others and if the `/etc/hosts` file or equivalent on other systems is not correctly set up then the computer can't reach localhost. I have actually experienced this on a number of computers. However `127.0.0.1` always works. And since you use the IP directly when setting up the server, it would be more consistent to do it here as well. Or do you want the hostname for some specific reason? Such as testing the `Host` field or such?
</blockquote></details>

No specific reason. I've updated it use `127.0.0.1`.

---

*[http/tests/custom_headers.rs, line 164 at r5](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7YThBpRvTVuiPQeCpG:-L7Z4u5G-4qjGwAv_-Kz:b-vz1wtb) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L164)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I have personally not seen "flag" used in this context I think. An object one keep track of in order to control another object somewhere I usually refer to as a "handle".

For me a flag is usually something that is sent in or returned out to indicate the activation of some feature.
</blockquote></details>

I've renamed it into `ServerHandle`, and moved it further below so that the service-specific stuff stay together. I considered making it into a normal struct with the port value set in it and with a `.port()` method. I didn't do it because it thought the ratio between the size of the change and the organization benefit of the refactor wouldn't be much worth it. We can always do that if we need to put more stuff in the handle. What do you think?

---

*[http/tests/custom_headers.rs, line 168 at r5](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7YT9Mnhrs06-hMMnEJ:-L7YuRxNEFMp6fyZhvGZ:bpng7kl) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L168)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Would it not work equally well without this `Drop` implementation? Since when `ServerShutdownFlag` is dropped the `Sender` will automatically be dropped and thus closed anyway. So then the `Receiver` will return in an error since the channel is closed. Both would make `run_until` return? Just thinking we can simplify the implementation a bit maybe.
</blockquote></details>

That makes sense, yeah. I've updated it to ignore any errors as well.

---

*[http/tests/custom_headers.rs, line 180 at r5](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7YUlpg0dYuu8XKXOem:-L7Z0DTc7s9_NBgaYEIa:bjqhw55) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L180)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

I have never used tokio services. Is this the simplest way to just listen for a single request on a http server? Seems like a lot of boilerplate to define two types implementing two traits just for that. But looking at the `Http` documentation it looks like it.
</blockquote></details>

Yeah, the idea is that the `NewService` is used to create `Service` instances, but they both receive `&self` in their methods because they must be thread-safe. So for simple tests that require non-thread-safe state, it becomes a little harder. The main blocker for needing all this boilerplate was that I was using a `oneshot` channel to notify the result of the header check.

I changed it to use `mpsc`, but then I had to add some similar boilerplate code to clone the `mpsc::Sender` for each instanced service. I'm not sure which is better:

- guaranteeing it only gets checked once (using `oneshot`);
- or having thread-safe code (using `mpsc`).

I'm open to suggestions and new ideas.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
Reviewed 1 of 3 files at r3.
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[http/tests/custom_headers.rs, line 164 at r5](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7YThBpRvTVuiPQeCpG:-L7ZIofk6xgMCufDLGQP:bkkkly4) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L164)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

I've renamed it into `ServerHandle`, and moved it further below so that the service-specific stuff stay together. I considered making it into a normal struct with the port value set in it and with a `.port()` method. I didn't do it because it thought the ratio between the size of the change and the organization benefit of the refactor wouldn't be much worth it. We can always do that if we need to put more stuff in the handle. What do you think?
</blockquote></details>

Now you don't need to put the oneshot channel in an `Option` any longer. Since it's never taken out.

---

*[http/tests/custom_headers.rs, line 180 at r5](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7YUlpg0dYuu8XKXOem:-L7ZJBXfJ_B7r7f-uy3G:bwfrx9c) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L180)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Yeah, the idea is that the `NewService` is used to create `Service` instances, but they both receive `&self` in their methods because they must be thread-safe. So for simple tests that require non-thread-safe state, it becomes a little harder. The main blocker for needing all this boilerplate was that I was using a `oneshot` channel to notify the result of the header check.

I changed it to use `mpsc`, but then I had to add some similar boilerplate code to clone the `mpsc::Sender` for each instanced service. I'm not sure which is better:

- guaranteeing it only gets checked once (using `oneshot`);
- or having thread-safe code (using `mpsc`).

I'm open to suggestions and new ideas.
</blockquote></details>

It would be more generally useful if the server could send all incoming requests over an `mpsc` channel. Then the mock server can be used to check anything back at the thread running the tests. Then asserts can fail with better error messages than just that the `check_result` was false. If you feel like implementing that.

That way we don't really have a `CheckService` rather just a `ForwardToChannel` service or similar.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[http/tests/custom_headers.rs, line 164 at r5](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7YThBpRvTVuiPQeCpG:-L7ZKMij-4fmmt83vjc7:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L164)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Now you don't need to put the oneshot channel in an `Option` any longer. Since it's never taken out.
</blockquote></details>

Done.

---

*[http/tests/custom_headers.rs, line 180 at r5](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7YUlpg0dYuu8XKXOem:-L7ZM6nVA_LkYpCjMOwe:bmi1bhu) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/94fdfc91c696c1dccf632fc9d5aae4417b4ba9fd/http/tests/custom_headers.rs#L180)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

It would be more generally useful if the server could send all incoming requests over an `mpsc` channel. Then the mock server can be used to check anything back at the thread running the tests. Then asserts can fail with better error messages than just that the `check_result` was false. If you feel like implementing that.

That way we don't really have a `CheckService` rather just a `ForwardToChannel` service or similar.
</blockquote></details>

Good idea! That reduced the code a lot! Thanks :)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/tests/custom_headers.rs, line 153 at r6](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7Z_ARDFsJbawfTevJJ:-L7Z_AREomjKAplHgdLo:buw11be) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/5e3d74c24fe1a5468c9b55c2b7159a9dfbccbb29/http/tests/custom_headers.rs#L153)):*
> ```Rust
>     fn call(&self, request: Request) -> Self::Future {
>         let mut sender = self.sender.clone();
>         let _ = sender.try_send(request);
> ```

Is there any reason these must be `futures::sync::mpsc` and not just `std::sync::mpsc`? Then you would not need to clone it, could just send directly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/tests/custom_headers.rs, line 153 at r6](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-L7Z_ARDFsJbawfTevJJ:-L7ZaxRslvjeGlfWg1nj:bcbgpbk) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/5e3d74c24fe1a5468c9b55c2b7159a9dfbccbb29/http/tests/custom_headers.rs#L153)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Is there any reason these must be `futures::sync::mpsc` and not just `std::sync::mpsc`? Then you would not need to clone it, could just send directly.
</blockquote></details>

Would also allow the handy `recv_timeout` on the other end.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

jvff commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[http/tests/custom_headers.rs, line 153 at r6](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32#-L7Z_ARDFsJbawfTevJJ:-L7ZhU_Z8a4VQwINzgKO:b-896fix) ([raw file](https://github.com/mullvad/jsonrpc-client-rs/blob/5e3d74c24fe1a5468c9b55c2b7159a9dfbccbb29/http/tests/custom_headers.rs#L153)):*
<details><summary><i>Previously, faern (Linus Färnstrand) wrote…</i></summary><blockquote>

Would also allow the handy `recv_timeout` on the other end.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/mullvad/jsonrpc-client-rs/32)*
<!-- Sent from Reviewable.io -->

faern commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mullvad/jsonrpc-client-rs/32#-:-L7ZnrI1nS9J1rH1iR0u:bnfp4nl)*
<!-- Sent from Reviewable.io -->
